runtime/compiler/trj9/optimizer/JProfilingRecompLoopTest.cpp

```diff
@@ -0,0 +1,214 @@
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+#include "JProfilingRecompLoopTest.hpp"
+#include "il/Block.hpp"
+#include "infra/Cfg.hpp"
+#include "infra/TRCfgEdge.hpp"
+#include "infra/Checklist.hpp"
+#include "infra/ILWalk.hpp"
+#include "infra/List.hpp"
+#include "optimizer/Optimization_inlines.hpp"
+#include "il/Node_inlines.hpp"
+#include "infra/Checklist.hpp"             // for TR::NodeChecklist
+#include "ras/DebugCounter.hpp"
+#include "control/Recompilation.hpp"              // for TR_Recompilation, etc
+#include "control/RecompilationInfo.hpp"              // for TR_Recompilation, etc
+#include "codegen/CodeGenerator.hpp"
+#include "optimizer/TransformUtil.hpp"            // for TransformUtil
+
+
+int32_t TR_JProfilingRecompLoopTest::recompileThreshold = 4096;
+
+
+int32_t TR_JProfilingRecompLoopTest::getLoopNestingLevel(int32_t nestingFactor)
+   {
+   if (nestingFactor == 1)
+      nestingFactor = 0;
+   
+   switch(nestingFactor)
+      {
+      case 0: return 0;
+      case 10: return 1;
+      case 100: return 2;
+      case 1000: return 3;
+      case 10000: return 4;
+      default: return 5;
+      }
+   }
+
+int32_t
+TR_JProfilingRecompLoopTest::perform()
+   {
+   if (comp()->getOption(TR_EnableJProfiling))
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has been enabled, run JProfiling\n");
+      }
+   else if (comp()->getProfilingMode() == JProfiling)
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has been enabled for profiling compilations, run JProfilingRecompLoopTest\n");
+      }
+   else
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has not been enabled, skip JProfilingRecompLoopTest\n");
+      return 0;
+      }
+   RecompilationTestLocations testLocations(RecompilationTestLocationAllocator(comp()->trMemory()->currentStackRegion()));
+   TR::TreeTop *cursor = comp()->getStartTree();
+   TR::CFG *cfg = comp()->getFlowGraph();
+
+   while (cursor)
+      {
+      if (cursor->getNode()->getOpCodeValue() == TR::asynccheck && cursor->getEnclosingBlock()->getStructureOf()->getContainingLoop() != NULL)
```

Comment:
getEnclosing block is going to walk the trees - you should track the block as you walk the trees to save having to find the block everytime. The asynccheck my not be in the loop directly - you could have another subgraph node so every asynccheck needs to compute its loop nesting depth.